src/Workspaces/Core/Portable/Diagnostics/DiagnosticAnalysisResult.cs

```diff
@@ -320,7 +320,8 @@ private static void VerifyDocumentMap(Project project, ImmutableDictionary<Docum
         {
             foreach (var documentId in map.Keys)
             {
-                Debug.Assert(project.GetTextDocument(documentId)?.SupportsDiagnostics() == true);
+                var textDocument = project.GetTextDocument(documentId) ?? project.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(documentId);
```

Comment:
So my usual concern about the safety of TryGetSourceGeneratedDocumentForAlreadyGeneratedId applies and how we know it won't return null here, but this method confuses me even before you changed it. The map can have diagnostics for stuff that already doesn't have documents anymore, even before we invented source generators?

Reply:
⏱️ Updating the comment to clarify TryGetSourceGeneratedDocumentForAlreadyGeneratedId is desired for the underlying purpose of this assertion (we are asserting that the document with this ID is already known to exist in the specific project instance at the time this method is called).

Reply:
➡️ Now updated

Reply:
Still a bit confused: should this be asserting (generally) that textDocument is not null, or something? This isn't new to your PR, but it seems like if that Try... always fails this will keep marching on, which might not be good? This is less a concern about your change and more a concern that I can't figure out what this was doing originally. @mavasani maybe you have some idea? "We're going to assert this document supports documents, unless it didn't exist entirely" seems like an odd assertion to make.

Reply:
> but it seems like if that Try... always fails this will keep marching on

If the `Try` fails, we'll get an assertion failure in test builds. The method doesn't run in any production scenario.

Reply:
Oh I misread that entirely; so yeah, sorry for the confusion here. I don't know why the original one used ? at all versus just asserting the result was non-null first or something.